README.md
# WhertTeX
Mindustry mod aiming to add stuff fitting in vanilla style well, while completely changing some game aspects.

## FAQ
Q: How do i set a firing spot for a Vertex cannon?
A: Select the cannon and then tap the desired tile TWICE. Please note that selection does not stick to units - you have to select tiles/blocks. Cannon won't attack your own buildings, so feel free to select them - Vertex will act as protection.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.0] - 2020-02-06
### Added
- Omnidirectional versions of item and fluid bridges
- Thermal centrifuge, larger, better version of separator.
## [0.1.1] - 2020-02-06
### Fixed
- Bug with bridge conduit power consumption
### Changed
- Thermal centrifuge rebalance

## [0.2] - 2020-02-06
### Added
- Discharge and Jetstream, advanced versions of Arc and Wave, respectively
- Compatability with build 104
### Changed
- Thermal centrifuge rebalance (again)

## [0.3] - 2020-02-26
### Added
- Vertex, the powerful (when used wisely) artillery turret.
### Changed
- Discharge rebalane

## [0.4] - 2020-03-09
### Added
- Flash silicon smelter, upgraded version of silicon smelter
- Oil burner, an efficient way to produce power from oil
### Changed
- Vertex rebalance

## [0.4.1] - 2020-03-22
### Fixed
- Added flash silicon smelter sprite
- Actually added oil burner


scripts/blocks/vertex.js
const Vertex = extendContent(ArtilleryTurret, "vertex", {
    playerPlaced(tile) {
        this.spots[tile.x + "," + tile.y] = tile.pos();
    },
    
    configured(tile, player, value) {
        this.spots[tile.x + "," + tile.y] = value;
    },
    
    drawConfigure(tile){
        Draw.color(Pal.accent);
        Lines.stroke(1.0);
        Lines.dashCircle(tile.drawx(), tile.drawy(), this.range);
        
        if(this.spots[tile.x + "," + tile.y] === undefined) {
            this.spots[tile.x + "," + tile.y] = tile.pos();
        }
        var spotTile = Vars.world.tile(this.spots[tile.x + "," + tile.y]);
        var size = 6 + Mathf.absin(Time.time(), 4, 2)
        
        var dx = tile.drawx() - spotTile.drawx();
        var dy = tile.drawy() - spotTile.drawy();
        var turretCircleRadius = (tile.block().size / 2 + 1) * Vars.tilesize + Mathf.absin(Time.time(), 6, 1) - 2
        var isTurretOutOfItsSpot = Math.pow(dx, 2) + Math.pow(dy, 2) >= Math.pow(this.spotRadius + turretCircleRadius, 2)
        //Calculate coords of the line connecting turret with firing spot
        
        if (isTurretOutOfItsSpot) {
            var normalisationQuotient = Math.sqrt(Math.pow(dx, 2) + Math.pow(dy, 2));
            var startX = tile.drawx() - dx / normalisationQuotient * turretCircleRadius;
            var startY = tile.drawy() - dy / normalisationQuotient * turretCircleRadius;
            var endX = spotTile.drawx() + dx / normalisationQuotient * this.spotRadius;
            var endY = spotTile.drawy() + dy / normalisationQuotient * this.spotRadius;
        }
        
        Lines.stroke(3.0);
        Draw.color(Pal.gray);
        if(isTurretOutOfItsSpot) Lines.dashLine(
            startX,
            startY,
            endX,
            endY,
            Math.sqrt((Math.pow(dx, 2) + Math.pow(dy, 2)) - turretCircleRadius - this.spotRadius) / 8
        )
        
        Lines.line(
            spotTile.drawx() - size, 
            spotTile.drawy() - size,
            spotTile.drawx() + size, 
            spotTile.drawy() + size
        )
        Lines.line(
            spotTile.drawx() - size, 
            spotTile.drawy() + size,
            spotTile.drawx() + size, 
            spotTile.drawy() - size
        )
     
        Lines.stroke(1.0);
        Draw.color(tile.getTeam().color);
        if(isTurretOutOfItsSpot) Lines.dashLine(
            startX,
            startY,
            endX,
            endY,
            Math.sqrt((Math.pow(dx, 2) + Math.pow(dy, 2)) - turretCircleRadius - this.spotRadius) / 8
        )
        
        Lines.line(
            spotTile.drawx() - size, 
            spotTile.drawy() - size,
            spotTile.drawx() + size, 
            spotTile.drawy() + size
        )
        Lines.line(
            spotTile.drawx() - size, 
            spotTile.drawy() + size,
            spotTile.drawx() + size, 
            spotTile.drawy() - size
        )
        
        Drawf.dashCircle(spotTile.drawx(), spotTile.drawy(), this.spotRadius, tile.getTeam().color);
        Drawf.circles(tile.drawx(), tile.drawy(), turretCircleRadius, tile.getTeam().color);
        
        Draw.reset();
    },
    
    drawSelect(tile) {
        // Only draw firing spot
        Lines.stroke(1.0);
        
        if(this.spots[tile.x + "," + tile.y] === undefined) {
            this.spots[tile.x + "," + tile.y] = tile.pos();
        }
        var spotTile = Vars.world.tile(this.spots[tile.x + "," + tile.y]);
        var size = 6 + Mathf.absin(Time.time(), 4, 2)
        
        Lines.stroke(3.0);
        Draw.color(Pal.gray);
        Lines.line(
            spotTile.drawx() - size, 
            spotTile.drawy() - size,
            spotTile.drawx() + size, 
            spotTile.drawy() + size
        )
        Lines.line(
            spotTile.drawx() - size, 
            spotTile.drawy() + size,
            spotTile.drawx() + size, 
            spotTile.drawy() - size
        )
        
        Lines.stroke(1.0);
        Draw.color(tile.getTeam().color);
        Lines.line(
            spotTile.drawx() - size, 
            spotTile.drawy() - size,
            spotTile.drawx() + size, 
            spotTile.drawy() + size
        )
        Lines.line(
            spotTile.drawx() - size, 
            spotTile.drawy() + size,
            spotTile.drawx() + size, 
            spotTile.drawy() - size
        )
        
        Drawf.dashCircle(spotTile.drawx(), spotTile.drawy(), this.spotRadius, tile.getTeam().color);
        
        Draw.reset();
    },
    
    onConfigureTileTapped(tile, other) {
        if(Math.pow(other.x - tile.x, 2) + Math.pow(other.y - tile.y, 2) > Math.pow((this.range - this.spotRadius) / Vars.tilesize, 2)) return true;
        if (this.spots[tile.x + "," + tile.y] === other.pos()) return true;
        tile.configure(other.pos());
        return false;
    },
    
    findTarget(tile) {
        entity = tile.ent();
        
        if(this.spots[tile.x + "," + tile.y] === undefined) {
            this.spots[tile.x + "," + tile.y] = tile.pos();
        }
        var spotTile = Vars.world.tile(this.spots[tile.x + "," + tile.y]);
        
        if (spotTile.block() != Blocks.air && spotTile.getTeam() != tile.getTeam()) entity.target = spotTile 
        else entity.target = Units.closestTarget(tile.getTeam(), spotTile.drawx(), spotTile.drawy(), this.spotRadius, boolf((e) => {return !e.isDead() && !e.isFlying()}));
    }
});

Vertex.spotRadius = 96;
Vertex.spots = {};


scripts/omnidirectional-bridge-conveyor.js
const omnidirectionalBridgeConveyor = extendContent(BufferedItemBridge, "omnidirectional-bridge-conveyor", {
    linkValid(tile, other, checkDouble){
        if(other == null || tile == null || other == tile) return false;
        if(Math.pow(other.x - tile.x, 2) + Math.pow(other.y - tile.y, 2) > Math.pow(this.range + 0.5, 2)) return false;
        return other.block() == this && (!checkDouble || other.ent().link != tile.pos());
    },
    drawPlace(x, y, rotation, valid) {
        Draw.color(Pal.placing);
        Lines.stroke(1.0);
        Lines.dashCircle(x * Vars.tilesize, y * Vars.tilesize, this.range * Vars.tilesize);

        
        link = this.findLink(x, y);
        if (link != null) {
            Lines.dashLine(x * Vars.tilesize, y * Vars.tilesize, link.drawx(), link.drawy(), this.range);
            const angle = Math.atan2(y - link.y, x - link.x);
            Draw.rect("bridge-arrow", 
            (link.x +  Math.cos(angle)) * Vars.tilesize, 
            (link.y + Math.sin(angle)) * Vars.tilesize, 
            angle * 180 / 3.14);
        }
        
        Draw.reset();
    },
    drawConfigure(tile) {
        var entity = tile.ent();
        
        Draw.color(Pal.placing);
        Lines.stroke(1.0);
        Lines.dashCircle(tile.x * Vars.tilesize, tile.y * Vars.tilesize, this.range * Vars.tilesize);

        Draw.reset();

        Draw.color(Pal.accent);
        Lines.stroke(1.0);
        Lines.square(tile.drawx(), tile.drawy(),
        tile.block().size * Vars.tilesize / 2.0 + 1.0);

        Geometry.circle(tile.x, tile.y, this.range + 1, new Intc2(){get: (x, y) => {
            other = Vars.world.ltile(x, y);
            if(this.linkValid(tile, other)) {
                var linked = other.pos() == entity.link;
                Draw.color(linked ? Pal.place : Pal.breakInvalid);

                Lines.square(other.drawx(), other.drawy(),
                other.block().size * Vars.tilesize / 2.0 + 1.0 + (linked ? 0.0 : Mathf.absin(Time.time(), 4.0, 1.0)));
            }
            }}
        );
        Draw.reset();
    },
    drawLayer(tile) {
        var entity = tile.ent();

        var other = Vars.world.tile(entity.link);
        if(!this.linkValid(tile, other)) return;

        var opacity = Core.settings.getInt("bridgeopacity") / 100.0;
        if(Mathf.zero(opacity)) return;
                                                    
        var uptime = Vars.state.isEditor() ? 1.0 : entity.uptime;
        const dx = other.worldx() - tile.worldx()
        const dy = other.worldy() - tile.worldy()
        var ex = dx * uptime
        var ey = dy * uptime
        
        Draw.color(Color.white, Color.black, Mathf.absin(Time.time(), 6.0, 0.07));
        Draw.alpha(Math.max(entity.uptime, 0.25) * opacity);
        
        const angle = Math.atan2(other.worldy() - tile.worldy(), other.worldx() - tile.worldx());
                                                    
        Draw.rect(this.endRegion, tile.worldx(), tile.worldy(), 90 + angle * 180 / 3.14);
        Draw.rect(this.endRegion, tile.worldx() + ex, tile.worldy() + ey, 270 + angle * 180 / 3.14);
        
        Lines.stroke(8.0);
        Lines.line(this.bridgeRegion,
        tile.worldx(),
        tile.worldy(),
        tile.worldx() + ex,
        tile.worldy() + ey, CapStyle.none, -Vars.tilesize / 2.0);
        
        const arrows = Math.floor(Math.sqrt(Math.pow(other.x - tile.x, 2) + Math.pow(other.y - tile.y, 2)) * Vars.tilesize / 6) - 1
        for(var a = 0; a != arrows; ++a) {
            const ox = dx / arrows * 0.666; // Like we skip 2/3 of arrow
            const oy = dy / arrows * 0.666;
            Draw.alpha(Mathf.absin(a / arrows - entity.time / 100.0, 0.1, 1) * uptime * opacity);
            Draw.rect(this.arrowRegion,
            tile.worldx() + ox + (dx - ox) * a / arrows * uptime,
            tile.worldy() + oy + (dy - oy) * a / arrows * uptime, angle * 180 / 3.14);
        }
        
        Draw.reset();
    },
    canDump(tile) {
        // Omnidirectional bridges dump items to all directions
        return true;
    },
    acceptItem(item, tile, source) {
        if(tile.getTeam() != source.getTeam()) return false;

        var entity = tile.ent();
        var other = Vars.world.tile(entity.link);

        if(this.linkValid(tile, other)) return tile.entity.items.total() < this.bufferCapacity;
        return source.block() instanceof ItemBridge && source.ent().link == tile.pos() && tile.entity.items.total() < this.bufferCapacity;

       
    },
});


scripts/blocks/omnidirectional-bridge-conduit.js
const omnidirectionalBridgeConduit = extendContent(LiquidExtendingBridge, "omnidirectional-bridge-conduit", {
    linkValid(tile, other, checkDouble){
        if(other == null || tile == null || other == tile) return false;
        if(Math.pow(other.x - tile.x, 2) + Math.pow(other.y - tile.y, 2) > Math.pow(this.range + 0.5, 2)) return false;
        return other.block() == this && (!checkDouble || other.ent().link != tile.pos());
    },
    drawPlace(x, y, rotation, valid) {
        Draw.color(Pal.placing);
        Lines.stroke(1.0);
        Lines.dashCircle(x * Vars.tilesize, y * Vars.tilesize, this.range * Vars.tilesize);

        
        link = this.findLink(x, y);
        if (link != null) {
            Lines.dashLine(x * Vars.tilesize, y * Vars.tilesize, link.drawx(), link.drawy(), this.range);
            const angle = Math.atan2(y - link.y, x - link.x);
            Draw.rect("bridge-arrow", 
            (link.x +  Math.cos(angle)) * Vars.tilesize, 
            (link.y + Math.sin(angle)) * Vars.tilesize, 
            angle * 180 / 3.14);
        }
        
        Draw.reset();
    },
    drawConfigure(tile) {
        var entity = tile.ent();
        
        Draw.color(Pal.placing);
        Lines.stroke(1.0);
        Lines.dashCircle(tile.x * Vars.tilesize, tile.y * Vars.tilesize, this.range * Vars.tilesize);

        Draw.reset();

        Draw.color(Pal.accent);
        Lines.stroke(1.0);
        Lines.square(tile.drawx(), tile.drawy(),
        tile.block().size * Vars.tilesize / 2.0 + 1.0);

        Geometry.circle(tile.x, tile.y, this.range + 1, new Intc2(){get: (x, y) => {
            other = Vars.world.ltile(x, y);
            if(this.linkValid(tile, other)) {
                var linked = other.pos() == entity.link;
                Draw.color(linked ? Pal.place : Pal.breakInvalid);

                Lines.square(other.drawx(), other.drawy(),
                other.block().size * Vars.tilesize / 2.0 + 1.0 + (linked ? 0.0 : Mathf.absin(Time.time(), 4.0, 1.0)));
            }
            }}
        );
        Draw.reset();
    },
    drawLayer(tile) {
        var entity = tile.ent();

        var other = Vars.world.tile(entity.link);
        if(!this.linkValid(tile, other)) return;

        var opacity = Core.settings.getInt("bridgeopacity") / 100.0;
        if(Mathf.zero(opacity)) return;
                                                    
        var uptime = Vars.state.isEditor() ? 1.0 : entity.uptime;
        const dx = other.worldx() - tile.worldx()
        const dy = other.worldy() - tile.worldy()
        var ex = dx * uptime
        var ey = dy * uptime
        
        Draw.color(Color.white, Color.black, Mathf.absin(Time.time(), 6.0, 0.07));
        Draw.alpha(Math.max(entity.uptime, 0.25) * opacity);
        
        const angle = Math.atan2(other.worldy() - tile.worldy(), other.worldx() - tile.worldx());
                                                    
        Draw.rect(this.endRegion, tile.worldx(), tile.worldy(), 90 + angle * 180 / 3.14);
        Draw.rect(this.endRegion, tile.worldx() + ex, tile.worldy() + ey, 270 + angle * 180 / 3.14);
        
        Lines.stroke(8.0);
        Lines.line(this.bridgeRegion,
        tile.worldx(),
        tile.worldy(),
        tile.worldx() + ex,
        tile.worldy() + ey, CapStyle.none, -Vars.tilesize / 2.0);
        
        const arrows = Math.floor(Math.sqrt(Math.pow(other.x - tile.x, 2) + Math.pow(other.y - tile.y, 2)) * Vars.tilesize / 6) - 1
        for(var a = 0; a != arrows; ++a) {
            const ox = dx / arrows * 0.666; // Like we skip 2/3 of arrow
            const oy = dy / arrows * 0.666;
            Draw.alpha(Mathf.absin(a / arrows - entity.time / 100.0, 0.1, 1) * uptime * opacity);
            Draw.rect(this.arrowRegion,
            tile.worldx() + ox + (dx - ox) * a / arrows * uptime,
            tile.worldy() + oy + (dy - oy) * a / arrows * uptime, angle * 180 / 3.14);
        }
        
        Draw.reset();
    },
    canDumpLiquid(tile, to, liquid) {
        // Omnidirectional bridges dump liquid to all directions
        return true;
    },
    acceptItem(item, tile, source) {
        return false;
    },
    acceptLiquid(tile, source, liquid, amount) {
       if(tile.getTeam() != source.getTeam() || !this.hasLiquids) return false;

        entity = tile.ent();
        other = Vars.world.tile(entity.link);

        if(this.linkValid(tile, other)) return tile.entity.liquids.get(liquid) + amount < this.liquidCapacity && (tile.entity.liquids.current() == liquid || tile.entity.liquids.get(tile.entity.liquids.current()) < 0.2);
        return source.block() instanceof ItemBridge && source.ent().link == tile.pos();
    }
});
